### PR TITLE
 PROBE_CALIBRATE command fails

### PR DESCRIPTION
**Pull request by GadgetAngel**
I did the following:

```BASH
cd /home/pi
git clone https://github.com/voidtrance/voron-klipper-extensions
./voron-klipper-extensions/install-extensions.sh
```
The `install-extensions.sh`` script produces the following error:
```
'Permission Denied' error 
```

Permission_denied

The ``install-extensions.sh``  script is a regular file without the ``execute bit``  The file permissions is ``644`` when it needs to be ``755``.

Here is a URL to help you setup the GitHub repo so that the ``install-extensions.sh``  script file will keep the state of the ``execute`` bit after it is uploaded to your GitHub Repo.  URL:  [How to commit executable shell scripts with Git on Windows](https://www.augmentedmind.de/2020/05/10/commit-executable-shell-scripts-with-git/)

100644 indicates that it is a regular file without the execute bit

After I fixed the ``execute`` bit (I know you can run ``chmod +x ~/voron-klipper-extensions/install-extensions.sh`` command, but this issue can be overcome so the user does not need to do that extra step), the script file ended up coping the whole GitHub directory contents to ``/home/pi/klipper/klippy/extras/gcode_shell_command.py`` and ``/home/pi/klipper/klippy/extras/settling_probe.py`` which are directories not files. See the below picture:

endup with gcode_shell_command pyfolder

After examining the ``install-extensions.sh``  script file, I found you had an extra ``s`` on the a variable in the ``for - do`` loop.  You had ``extensions`` where it needed to be ``extension``.  Why did the  ``for - do`` loop do a copy instead of a symbolic link??

Well I corrected the error in the variable name and removed the directory from my raspberry pi and re-cloned my version of your GitHub to test out the fix.

I also add a function ``restart_klipper`` so that once the files' symbolic links are created the install script will restart klipper so the extensions will be installed.

Now the install script works as it was intended.  

I now started testing the code.  Here are the results:
```
PROBE                                  #worked as expected
PROBE SETTLING_SAMPLE=0                #worked as expected
PROBE SETTLING_SAMPLE=1                #worked as expected

PROBE_ACCURACY                         #worked as expected
PROBE_ACCURACY SETTLING_SAMPLE=0       #worked as expected
PROBE_ACCURACY SETTLING_SAMPLE=1       #worked as expected

PROBE_CALIBRATE                        #worked as expected, after fix applied
PROBE_CALIBRATE SETTLING_SAMPLE=0      #worked as expected, after fix applied
PROBE_CALIBRATE SETTLING_SAMPLE=1      #worked as expected, after fix applied

QUAD_GANTRY_LEVEL                      #worked as expected
QUAD_GANTRY_LEVEL SETTLING_SAMPLE=0    #works but it ignores the command line parameter and defaults to the config setting instead
QUAD_GANTRY_LEVEL SETTLING_SAMPLE=1    #works but it ignores the command line parameter and defaults to the config setting instead

BED_MESH_CALIBRATE                     #worked as expected
BED_MESH_CALIBRATE SETTLING_SAMPLE=0   #works but it ignores the command line parameter and defaults to the config setting instead
BED_MESH_CALIBRATE SETTLING_SAMPLE=1   #works but it ignores the command line parameter and defaults to the config setting instead
```

I understand that the ``QUAD_GANTRY_LEVEL`` and ``BED_MESH_CALIBRATE`` are working as per your designed.

I would like a feature added to allow these command to accept the command line parameter ``SETTLING_SAMPLE`` if the printer the code is installed are is a printer with a flying gantry like the Voron2 

I did further testing on the auto z offset commands and here are the results:
```
PROBE_Z_ACCURACY                       # does not discard the first sample
CALIBRATE_Z                            # does not discard the first sample
```

I know that you did not design this extension to include ``Auto-Z-Offset`` but again I would request a feature be added.

I am also documented for others that read this post what the capabilities of this extension will give them and what to expect when they install it. 

Here is the output from the CALIBRATE_Z command:
```
CALIBRATE_Z
echo: off
echo: calibrating_z
probe: open
MagProbe: Already attached
probe: open
probe: open
MagProbe: Dock Probe
echo: off
echo: busy
echo: off
echo: complete
probe: TRIGGERED
probe at 202.000,303.500 is z=2.171432
probe at 202.000,303.500 is z=2.163932
probe at 202.000,303.500 is z=2.163932
probe at 202.000,303.500 is z=2.163932
probe: TRIGGERED
MagProbe: Attach Probe
echo: off
echo: busy
echo: off
echo: complete
probe: open
probe at 207.000,278.500 is z=6.776432
probe at 207.000,278.500 is z=6.790182
probe at 207.000,278.500 is z=6.790182
probe at 207.000,278.500 is z=6.788932
probe at 150.000,125.000 is z=4.785182
probe at 150.000,125.000 is z=4.801432
probe at 150.000,125.000 is z=4.801432
probe at 150.000,125.000 is z=4.800182
Z-CALIBRATION: ENDSTOP=2.159 NOZZLE=2.164 SWITCH=6.790 PROBE=4.801 --> OFFSET=-0.062318
probe: open
MagProbe: Dock Probe
echo: off
echo: busy
echo: off
echo: complete
probe: TRIGGERED
probe: TRIGGERED
MagProbe: Already docked
probe: TRIGGERED
```

## Thank you for all the work you have done on this extension I will enjoy using it.

### CODE REVIEW

**Comment by voidtrance** [comment]
Thank you for the PR and fixes.

There is a lot in a single PR, so I'll try to address things one-by-one:

1. The lack of support for the the `SETTLING_SAMPLE` argument to `QUAD_GANTRY_LEVEL` and `BED_MESH_CALIBRATE` is intentional. While there is a very easy way to get all Z probing commands to accept the command line argument, it goes much beyond the purpose of this extension. As described, this extension is primarily for Voron printer (more specifically - Voron 2.4 printers), which only use the `QUAD_GANTRY_LEVEL` command. Adding the argument to Z probing commands will affect all printers, not just Vorons. I am not comfortable doing that as of yet. If there is interest in supporting other printers, I can revisit this.
2. I looking into making the extension work with Auto-Z calibration. Unfortunately, I am not sure whether this will be possible. The Auto-Z extension completely replicates the code from `probe.py` rather than using the `probe.ProbePointsHelper` class. In order to get Auto-Z to work as is, its code will have to change. We will either have to submit a feature request to add this directly to Auto-Z or change it so it uses `probe.ProbePointsHelper`.

As for this PR, couple you please squash commits as appropriate? There are commits which fix the README and then commits that fix the first commit. Those are perfect for squashing into a single commit. I'll also provide some feedback in the individual commits.

**Review by voidtrance — COMMENTED** [review]
The commit message seems wrong. The only thing that this commit does is change the file permissions, yet the commit message says a lot more.

Also, could you please change the commit messages to follow the format:

> short subject line
>
> Longer description of the changes
> in this commit

File: install-extensions.sh [hunk 1]
```diff
@@ -26,6 +26,13 @@ function link_extension() {
     done
 }
 
+# Step 3: restarting Klipper
+function restart_klipper()
+{
+    echo "Restarting Klipper..."
+    sudo systemctl restart klipper
+}
```
Comment (voidtrance) [inline install-extensions.sh:34]: Before taking this, let me do some research into Moonraker. Their documentation seems to suggest that the `managed_services` setting should be used to let Moonraker know which services to restart. Therefore, there shouldn't be a need to restarted it as part of the script.

File: install-extensions.sh [hunk 2]
```diff
@@ -18,14 +18,28 @@ function check_klipper() {
     fi
 }
 
-# Step 2: Link extension to Klipper
-function link_extension() {
-    echo "Linking extensions to Klipper..."
-    for extension in ${EXTENSION_LIST}; do
-        ln -sf "${SRCDIR}/${extensions}" "${KLIPPER_PATH}/klippy/extras/${extension}"
```
Comment (voidtrance) [inline install-extensions.sh:25]: I don't think that we should have separate functions for each file. That will get harder to maintain if there are more extensions added.

I think the main issue is the spelling mistake. If you replace `${extensions}` with `${extension}`, it should only link individual files, not the whole directory.

File: install-extensions.sh [hunk 3]
```diff
@@ -29,7 +29,7 @@ function link_gcode_shell_command_extension()
 function link_settling_probe_command_extension()
 {
     echo "Linking extension to Klipper..."
-    ln -sf "${SRCDIR}/gcode_shell_command.py" "${KLIPPER_PATH}/klippy/extras/gcode_shell_command.py"
+    ln -sf "${SRCDIR}/settling_probe.py" "${KLIPPER_PATH}/klippy/extras/settling_probe.py"
```
Comment (voidtrance) [inline install-extensions.sh]: This commit is a perfect example of something that should have been squashed into a single commit. Also, it's a good example of a commit that needs it's commit message fixed.

Comment (GadgetAngel) [reply install-extensions.sh]: I found the mistakes and fixed them.  The last commit has this fixed.  I do not know how to "squish" a commit, do you want to teach me?

File: install-extensions.sh (hunk 2 above)
Comment (GadgetAngel) [reply install-extensions.sh:25]: This was fixed in my last commit

File: install-extensions.sh (hunk 1 above)
Comment (GadgetAngel) [reply install-extensions.sh:34]: The problem is when I was testing out the install script, I need to do it a couple of times to fix the issues I found.   I was not using moonraker to do that.

Moonraker will only come into play when you do a release of your Github repo which add a release tag to your GitHub rep and then moonraker can update if you make changes down the road.

Having the restart in the file does not hurt it, even if you have the moonraker option to bind Klipper to your extension, but it sure comes in handy for testing

**Comment by GadgetAngel** [comment]
Tried updating via moonraker today and got this message:

image

Plus your Repo is showing INVALID:

image


For people to update via moonraker you need to add a tag (version number) to your Github Repo and you do that by publishing a RELEASE of your Repo.

Here is a link on how to do a RELEASE and add a tag (version number) to your Repo.  Moonraker will not consider your Repo `pristine` until you add a tag

https://docs.github.com/en/repositories/releasing-projects-on-github/managing-releases-in-a-repository

**Comment by voidtrance** [comment]
According to the documentation, there shouldn't be a need for tags or releases:

```
channel: dev
#   The update channel.  The available value differs depending on the
#   "type" option.
#      type: git_repo - May be dev or beta.  The dev channel will update to
#                       the latest pushed commit, whereas the beta channel
#                       will update to the latest tagged commit.
#      type: zip      - May be be stable or beta.  When beta is specified
#                       "pre-release" updates are available.
#   The default is dev.
```
But then there could be some issue with Moonraker. I'll have to check.

**Comment by voidtrance** [comment]
I have to say that this pull request is starting to be a dumping ground for multiple issues. Please, don't overload it with multiple issue?

File: install-extensions.sh (hunk 1 above)
Comment (GadgetAngel) [reply install-extensions.sh:34]: For some information, having Klipper restart does not effect the ability for moonraker to update automatically.  moonraker will just delay the Klipper restart when it detects it during and update. That way the install occurs and the restart is placed at the end

**Comment by GadgetAngel** [comment]
> According to the documentation, there shouldn't be a need for tags or releases:
> 
> ```
> channel: dev
> #   The update channel.  The available value differs depending on the
> #   "type" option.
> #      type: git_repo - May be dev or beta.  The dev channel will update to
> #                       the latest pushed commit, whereas the beta channel
> #                       will update to the latest tagged commit.
> #      type: zip      - May be be stable or beta.  When beta is specified
> #                       "pre-release" updates are available.
> #   The default is dev.
> ```
> 
> But then there could be some issue with Moonraker. I'll have to check.
image

**Comment by GadgetAngel** [comment]
> I have to say that this pull request is starting to be a dumping ground for multiple issues. Please, don't overload it with multiple issue?

Just trying to help, I will stop

**Comment by voidtrance** [comment]
Don't get me wrong! There is no need to stop just file different issues for different bugs. You don't even need to submit pull requests if you don't feel like it. I just don't want this to become a convoluted discussion for multiple issues/bugs. That's going to get messy and hard to maintain/manage.

**Comment by GadgetAngel** [comment]
> Don't get me wrong! There is no need to stop. Just file different issues for different bugs. You don't even need to submit pull requests if you don't feel like it. I just don't want this to become a convoluted discussion for multiple issues/bugs. That's going to get messy and hard to maintain/manage.

I understand now

**Pull request closed without merging**